providers/base/tests/test_pipewire_utils.py

```diff
@@ -0,0 +1,634 @@
+        pt = PipewireTest()
+
+        # return Audio/Sink
+        self.assertEqual("Audio/Sink",
+                         pt.generate_pw_media_class("audio", "sink"))
+        self.assertEqual("Audio/Sink",
+                         pt.generate_pw_media_class("audio", "Sink"))
+        self.assertEqual("Audio/Sink",
+                         pt.generate_pw_media_class("Audio", "sink"))
+        self.assertEqual("Audio/Sink",
+                         pt.generate_pw_media_class("Audio", "Sink"))
+        self.assertEqual("Audio/Sink",
+                         pt.generate_pw_media_class("Audio", "Sinks"))
+
+        # return Audio/Source
+        self.assertEqual("Audio/Source",
+                         pt.generate_pw_media_class("audio", "source"))
+        self.assertEqual("Audio/Source",
+                         pt.generate_pw_media_class("audio", "Source"))
+        self.assertEqual("Audio/Source",
+                         pt.generate_pw_media_class("Audio", "source"))
+        self.assertEqual("Audio/Source",
+                         pt.generate_pw_media_class("Audio", "Source"))
+        self.assertEqual("Audio/Source",
+                         pt.generate_pw_media_class("Audio", "Sources"))
+
+        # return Video/Sink
+        self.assertEqual("Video/Sink",
+                         pt.generate_pw_media_class("video", "sink"))
+        self.assertEqual("Video/Sink",
+                         pt.generate_pw_media_class("video", "Sink"))
+        self.assertEqual("Video/Sink",
+                         pt.generate_pw_media_class("Video", "sink"))
+        self.assertEqual("Video/Sink",
+                         pt.generate_pw_media_class("Video", "Sink"))
+        self.assertEqual("Video/Sink",
+                         pt.generate_pw_media_class("Video", "Sinks"))
+
+        # return Video/Source
+        self.assertEqual("Video/Source",
+                         pt.generate_pw_media_class("video", "source"))
+        self.assertEqual("Video/Source",
+                         pt.generate_pw_media_class("video", "Source"))
+        self.assertEqual("Video/Source",
+                         pt.generate_pw_media_class("Video", "source"))
+        self.assertEqual("Video/Source",
+                         pt.generate_pw_media_class("Video", "Source"))
+        self.assertEqual("Video/Source",
+                         pt.generate_pw_media_class("Video", "Sources"))
+
+    def test_fail(self):
+        pt = PipewireTest()
+
+        # return UNKNOWN TYPE
+        wrong_type1 = "videog"
+        wrong_type2 = "xxxx"
+        self.assertEqual("UNKNOWN TYPE",
+                         pt.generate_pw_media_class(wrong_type1,
+                                                    "source"))
+        self.assertEqual("UNKNOWN TYPE",
+                         pt.generate_pw_media_class(wrong_type2,
+                                                    "sources"))
+
+        # example code for testing log output in python 3.10 and above
```

Comment:
Do you mean this unit test has different results based on the Python version?

Reply:
`assertLogs` couldn't work well on `python 3.8` and lower as mention in the beginning of the file.

```
# There are some issue with assertLog for python version lower than 3.10,
# https://github.com/python/cpython/commit/6fdfcec5b11f44f27aae3d53ddeb004150ae1f61
# Therefore, please don't add new test cases of assertLog.
```

Reply:
From the link you shared, it seems the issue has been resolved.

Reply:
the [test result](https://github.com/canonical/checkbox/actions/runs/6888107337/job/18736472019?pr=826) from github action shows it is not for python 3.8.